File: android/app/src/main/java/app/releaf/mobile/data/notebook/PageEntity.kt
/*
 * PageEntity.kt
 *
 * Room entity mirroring the `pages` table in
 * `design-system/migrations/v1_initial.sql`. A page is the authored leaf of
 * the notebook → chapter → page hierarchy and carries the note body that
 * FTS5 indexes (`fts_page_notes`).
 *
 * This file only models the page row itself. The sibling capture rows
 * (photos, voice notes, scans) live in the `captures` table, which will be
 * modeled in a later phase — today the page just stores its title + markdown
 * notes plus the conflict-resolution stub for sync.
 */

package app.releaf.mobile.data.notebook

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "pages",
    indices = [
        // Primary list-pattern: pages within a chapter, ordered by position.
        Index(value = ["chapter_id", "position"]),
        Index("project_id"),
        Index("updated_at"),
        Index("deleted_at"),
        Index("archived_at"),
    ],
)
data class PageEntity(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "chapter_id")
    val chapterId: String,

    @ColumnInfo(name = "project_id")
    val projectId: String? = null,

    @ColumnInfo(name = "template_id")
    val templateId: String? = null,

    /** Nullable — the UI falls back to a derived default if absent. */
    @ColumnInfo(name = "title")
    val title: String? = null,

    /** Canonical CommonMark. Empty string is valid. */
    @ColumnInfo(name = "notes", defaultValue = "''")
    val notes: String = "",

    /** JSON array of Contact. Always a JSON string, never null. */
    @ColumnInfo(name = "contacts", defaultValue = "'[]'")
    val contacts: String = "[]",

    /** JSON array of Location. Always a JSON string, never null. */
    @ColumnInfo(name = "locations", defaultValue = "'[]'")
    val locations: String = "[]",

    /**
     * JSON array of TodoItem. Supplementary to any checklist items the
     * user writes inline in the notes body — these are first-class
     * tasks surfaced in the TODOS section of the page editor.
     */
    @ColumnInfo(name = "todos", defaultValue = "'[]'")
    val todos: String = "[]",

    /**
     * JSON array of Attachment (photos + scans). Kept on-page as JSON
     * rather than a separate `captures` table for now — see
     * PageAttachments.kt for the trade-off.
     */
    @ColumnInfo(name = "attachments", defaultValue = "'[]'")
    val attachments: String = "[]",

    /**
     * JSON array of Stroke (freehand pen/highlighter strokes overlaying the
     * notes body). Coordinates are pinned — strokes don't reflow when text
     * edits push content around. See Stroke in PageAttachments.kt.
     *
     * Deprecated path now: when `sub_pages` carries content, it owns the
     * stroke list per sub-page. This column is kept in sync with the first
     * sub-page's strokes for back-compat and for rows that pre-date v3.
     */
    @ColumnInfo(name = "sketch_strokes", defaultValue = "'[]'")
    val sketchStrokes: String = "[]",

    /**
     * JSON array of SubPage — the horizontal sub-pages the user swipes
     * between inside this page. Source of truth for notes + strokes
     * once populated. Empty string `'[]'` means "this is a legacy row
     * with content in the flat `notes` + `sketch_strokes` columns" —
     * the VM synthesizes a single sub-page from them on load.
     */
    @ColumnInfo(name = "sub_pages", defaultValue = "'[]'")
    val subPages: String = "[]",

    @ColumnInfo(name = "position", defaultValue = "1024")
    val position: Long = 1024L,

    /** JSON: `{local_notes, remote_notes, remote_updated_at}` or null. */
    @ColumnInfo(name = "conflict_stub")
    val conflictStub: String? = null,

    @ColumnInfo(name = "drive_file_id")
    val driveFileId: String? = null,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String,

    @ColumnInfo(name = "dirty", defaultValue = "1")
    val dirty: Boolean = true,

    @ColumnInfo(name = "deleted_at")
    val deletedAt: String? = null,

    /**
     * JSON array of free-form tag strings. Added in v17 — never
     * null at the row level (Migration16To17 backfills `'[]'` and
     * the column default keeps it that way for new rows). Drive
     * surfaces tag pills under the page title.
     */
    @ColumnInfo(name = "tags", defaultValue = "'[]'")
    val tags: String = "[]",

    /**
     * JSON array of `{id, body, createdAt}` objects — the typed
     * `Note` collection on the Drive-facing `Page` model. Added in
     * v18 so each note round-trips with its own id + timestamp
     * instead of being collapsed into a single markdown blob in
     * `notes`. The `notes` column still carries the joined
     * markdown for FTS indexing; this column is the source of
     * truth for note identity.
     */
    @ColumnInfo(name = "page_notes_json", defaultValue = "'[]'")
    val pageNotesJson: String = "[]",

    /**
     * Soft-archive timestamp (separate from `deletedAt`). Added in
     * v17 — non-nil means the page lives in the archive bin and is
     * filtered out of the active page list. Independent of
     * `deletedAt` (true tombstone).
     */
    @ColumnInfo(name = "archived_at")
    val archivedAt: String? = null,
)

File: ios/Releaf/Data/Notebook/PageEntity.swift
/*
 * PageEntity.swift
 *
 * GRDB record for the `pages` table — the authored leaf of the
 * notebook → chapter → page hierarchy. Carries the markdown body that
 * `fts_page_notes` indexes, plus the four feature-section JSON columns
 * (contacts, locations, todos, attachments) that match NotepadEntry.
 *
 * `Page` is claimed by the Drive-facing domain type in
 * `Data/Domain/Models.swift`, so this struct uses the `Entity` suffix
 * to avoid a duplicate declaration in the `ReleafData` module.
 */

import Foundation
import GRDB

public struct PageEntity: Codable, FetchableRecord, PersistableRecord,
                          Identifiable, Equatable, Sendable {

    public static let databaseTableName = "pages"

    public var id: String
    public var chapterId: String
    public var projectId: String?
    public var templateId: String?
    /// Nullable — the UI falls back to a derived default when absent.
    public var title: String?
    /// Canonical CommonMark. Empty string is valid.
    public var notes: String
    /// JSON array of `NotepadContact` (same shape as notepad). Never nil; `"[]"` default.
    public var contacts: String
    /// JSON array of `GeoLocation`. Never nil; `"[]"` default.
    public var locations: String
    /// JSON array of `NotepadTodo`. Never nil; `"[]"` default.
    public var todos: String
    /// JSON array of `Attachment` (photo/scan manifests). Never nil.
    public var attachments: String
    public var position: Int64
    /// JSON `{local_notes, remote_notes, remote_updated_at}` or nil.
    public var conflictStub: String?
    public var driveFileId: String?
    public var createdAt: String
    public var updatedAt: String
    public var dirty: Bool
    public var deletedAt: String?

    public enum CodingKeys: String, CodingKey {
        case id
        case chapterId    = "chapter_id"
        case projectId    = "project_id"
        case templateId   = "template_id"
        case title
        case notes
        case contacts
        case locations
        case todos
        case attachments
        case position
        case conflictStub = "conflict_stub"
        case driveFileId  = "drive_file_id"
        case createdAt    = "created_at"
        case updatedAt    = "updated_at"
        case dirty
        case deletedAt    = "deleted_at"
    }

    public init(
        id: String,
        chapterId: String,
        projectId: String? = nil,
        templateId: String? = nil,
        title: String? = nil,
        notes: String = "",
        contacts: String = "[]",
        locations: String = "[]",
        todos: String = "[]",
        attachments: String = "[]",
        position: Int64 = 1024,
        conflictStub: String? = nil,
        driveFileId: String? = nil,
        createdAt: String,
        updatedAt: String,
        dirty: Bool = true,
        deletedAt: String? = nil
    ) {
        self.id = id
        self.chapterId = chapterId
        self.projectId = projectId
        self.templateId = templateId
        self.title = title
        self.notes = notes
        self.contacts = contacts
        self.locations = locations
        self.todos = todos
        self.attachments = attachments
        self.position = position
        self.conflictStub = conflictStub
        self.driveFileId = driveFileId
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.dirty = dirty
        self.deletedAt = deletedAt
    }
}
